contrib/gazebo/gazebo_sawyer_env.py

```diff
@@ -0,0 +1,499 @@
+        else:
+            # should raise an unknown mode exception
+            raise ControlModeUnknownErr(
+                "Control mode %s is not known!" % self._control_mode)
+
+        self._action_space = Box(
+            np.concatenate((lower_bounds, np.array([0]))),
+            np.concatenate((upper_bounds, np.array([100]))))
+
+    def update_model_states(self, data):
+        self.model_states = data
+
+    def compute_reward(self, achieved_goal, goal, info=None):
+        # Compute distance between goal and the achieved goal.
+        d = goal_distance(achieved_goal, goal)
+        if self.reward_type == 'sparse':
+            return -(d > self.distance_threshold).astype(np.float32)
+        else:
+            return -d
+
+    # Implementation for rllab env functions
+    # -----------------------------------
+    @rate_limited(100)
+    def step(self, action):
+        """
+        Perform a step in gazebo. When end of episode
+        is reached, reset() should be called to reset the environment's internal state.
+        Input
+        -----
+        action : an action provided by the environment
+        Outputs
+        -------
+        (observation, reward, done, info)
+        observation : agent's observation of the current environment
+        reward [Float] : amount of reward due to the previous action
+        done : a boolean, indicating whether the episode has ended
+        info : a dictionary containing other diagnostic information from the previous action
+        """
+        print("Step forwarding...")
+
+        if self._control_mode != 'position':
+            rospy.wait_for_service('/gazebo/unpause_physics')
+            try:
+                unpause = rospy.ServiceProxy('/gazebo/unpause_physics',
+                                             std_srvs.srv.Empty)
+                unpause()
+            except rospy.ServiceException as e:
+                print('/gazebo/unpause_physics service call failed')
+
+        # set joint angles
+        joint_angle_cmds = {
+            'right_j0': action[0],
+            'right_j1': action[1],
+            'right_j2': action[2],
+            'right_j3': action[3],
+            'right_j4': action[4],
+            'right_j5': action[5],
+            'right_j6': action[6]
+        }
+
+        if self._control_mode == 'position':
+            self._limb.set_joint_positions(joint_angle_cmds)
+        elif self._control_mode == 'velocity':
+            self._limb.set_joint_velocities(joint_angle_cmds)
+        elif self._control_mode == 'effort':
+            self._limb.set_joint_torques(joint_angle_cmds)
+
+        gripper_cmd = action[7]
+        self._gripper.set_position(gripper_cmd)
+
+        obs = self._get_obs()
+
+        if self._control_mode != "position":
+            rospy.wait_for_service('/gazebo/pause_physics')
+            try:
+                pause = rospy.ServiceProxy('/gazebo/pause_physics',
```

Comment:
please predefine all your ServiceProxys either on import or construction, and make them class or package members